mmda/types/annotation.py

```diff
@@ -42,54 +59,72 @@ def __init__(
             self,
             id: Optional[int] = None,
             doc: Optional['Document'] = None,
+            field: Optional[str] = None,
             metadata: Optional[Metadata] = None
     ):
         self.id = id
         self.doc = doc
+        self.field = field
         self.metadata = metadata if metadata else Metadata()
 
     @abstractmethod
     def to_json(self) -> Dict:
-        pass
+        raise NotImplementedError
 
     @classmethod
     @abstractmethod
     def from_json(cls, annotation_dict: Dict) -> "Annotation":
-        pass
+        raise NotImplementedError
+
+    @property
+    def name(self) -> Optional[AnnotationName]:
+        if self.field and self.id:
+            return AnnotationName(field=self.field, id=self.id)
+        else:
+            return None
 
-    def attach_doc(self, doc: "Document") -> None:
+    def _attach_doc(self, doc: "Document", field: str) -> None:
         if not self.doc:
             self.doc = doc
+            self.field = field
         else:
             raise AttributeError("This annotation already has an attached document")
 
-    # TODO[kylel] - comment explaining
-    def __getattr__(self, field: str) -> List["Annotation"]:
-        if self.doc is None:
-            raise ValueError("This annotation is not attached to a document")
+    def _get_siblings(self) -> List['Annotation']:
+        """This method gets all other objects sharing the same field as the current object.
+        Only works after a Document has been attached, which is how objects learn their `field`."""
+        if not self.doc:
+            raise AttributeError("This annotation does not have an attached document")
+        return self.doc.__getattribute__(self.field)
 
-        if field in self.doc.fields:
-            return self.doc.find_overlapping(self, field)
+    def __getattr__(self, field: str) -> List["Annotation"]:
+        """This method allows jumping from an object of one field to all overlapping
+        objects of another field. For example `page.tokens` jumps from a particular page
+        to all its intersecting tokens."""
+        if not self.doc:
+            raise AttributeError("This annotation does not have an attached document")
 
         if field in self.doc.fields:
             return self.doc.find_overlapping(self, field)
 
+        # TODO[kylel] - when does this ever get called? infinite loop?
         return self.__getattribute__(field)
 
 
-
 class BoxGroup(Annotation):
     def __init__(
             self,
             boxes: List[Box],
             id: Optional[int] = None,
             doc: Optional['Document'] = None,
+            field: Optional[str] = None,
             metadata: Optional[Metadata] = None,
```

Comment:
Not related to this PR specifically, but I think we should document these arguments in a docstring.

Reply:
yea we'll need that